Log subsuite name in wptrunner log

WeizhongX opened the pull request:
Doing this to allow devs to know what exact tests are running or failed.

The test from a subsuite will be logged as "subsuite:test", or just test for generic subsuite. Vendors can also override this to log the test name in a different format. For example chromium wants to log the test name as virtual/subsuite/test.

WeizhongX committed:
Log subsuite name in wptrunner log

Doing this to allow devs to know what exact tests are running or failed.

The test from a subsuite will be logged as "subsuite:test", or just test
for generic subsuite. Vendors can also override this to log the test name
in a different format. For example chromium wants to log the test name as
virtual/subsuite/test.

WeizhongX commented:
A test run in chromium can be found here: https://chromium-swarm.appspot.com/task?id=6405826148ee6210&w=true.

jonathan-j-lee commented on tools/third_party_modified/mozlog/mozlog/formatters/grouping.py:
```diff
@@ -163,8 +164,28 @@ def wrap_and_indent_lines(self, lines, indent):
             output += indent + u"\u2514 %s\n" % lines[-1]
         return output
 
+    def get_lines_for_unexpected_subtest_result(
```
Rather than duplicating what was `get_lines_for_unexpected_result`, how about make the first argument the "formatted test key" (i.e., it becomes the caller's responsibility to pass either `subtest_name` or `self.get_test_name_output(subsuite, test_name)`)?

WeizhongX replied:
Done, thanks

jonathan-j-lee commented on tools/third_party_modified/mozlog/mozlog/formatters/grouping.py:
```diff
@@ -150,7 +151,7 @@ def suite_start(self, data):
 
     def test_start(self, data):
-        self.running_tests[data["thread"]] = data["test"]
+        self.running_tests[data["thread"]] = (data["subsuite"], data["test"])
```
The RFC says the [`subsuite` key is optional](https://github.com/web-platform-tests/rfcs/blob/51538dcf1666eb2a4e6c52386bb49afe5441b3af/rfcs/subsuites.md?plain=1#L163). Therefore, should this (and elsewhere) be `data.get("subsuite")` (if you choose `None` to represent the default subsuite)? (Should also help the tests pass.)

WeizhongX replied:
Looking at the implementation in https://github.com/web-platform-tests/wpt/pull/39711, `tests_to_run` etc is keyed with `(subsuite_name, test_name)`, where subsuite_name is the empty string for generic suite. So the subsuite_name is mandatory even though we want it to be optional in the RFC.

There once had a discussion if we want to use the empty string or None to represent the default subsuite. Even though we (chromium) to prefer to use None, jgraham@ choose to use the empty string. There might have a technical reason behind this. For example it might be difficult to represent None in metadata file. I think we can do a follow up change in future to use None instead of the empty string.

Yes using `data.get("subsuite")` can help pass the unit tests. I will take a look to see if we should use `data.get("subsuite")` and add unit tests to test the new scenarios, or make change to existing unit tests.

jonathan-j-lee replied:
> So the subsuite_name is mandatory even though we want it to be optional in the RFC.

The example of `test_to_run` in #39711 is different, since that's an internal wptrunner data structure. OTOH, if we ever choose to reconcile this vendored `mozlog` version with the `mozbase` one, the formatters should gracefully accommodate callers that don't pass `subsuite`.

> I think we can do a follow up change in future to use None instead of the empty string.

I agree that `None` would be preferable to make the default subsuite's special status obvious.

jonathan-j-lee reviewed (approved):
LGTM with comments addressed

WeizhongX committed:
Updated per review comments

Removed duplicated functions: get_lines_for_unexpected_subtest_result
and get_lines_for_unexpected_result. Use data.get("subsuite") instead
to align with other part of coding style.

jonathan-j-lee reviewed (approved):
Still LGTM

WeizhongX commented:
@jgraham , can you pls take a look?

I planned to add some unit tests for the changes made here, but seems most of the change requires an interactive terminal, so decided to give it up. If you have any thoughts, pls let me know, and I will be happy to add that as a separate PR.
